main/midi/midiUtil.c

```diff
@@ -0,0 +1,396 @@
+    0x01062295, // +41 cents => 1.02397
+    0x0106495b, // +42 cents => 1.02456
+    0x01067027, // +43 cents => 1.02515
+    0x010696f8, // +44 cents => 1.02574
+    0x0106bdd0, // +45 cents => 1.02633
+    0x0106e4ad, // +46 cents => 1.02693
+    0x01070b8f, // +47 cents => 1.02752
+    0x01073278, // +48 cents => 1.02811
+    0x01075966, // +49 cents => 1.02871
+    0x0107805a, // +50 cents => 1.02930
+    0x0107a754, // +51 cents => 1.02990
+    0x0107ce53, // +52 cents => 1.03049
+    0x0107f558, // +53 cents => 1.03109
+    0x01081c64, // +54 cents => 1.03168
+    0x01084374, // +55 cents => 1.03228
+    0x01086a8b, // +56 cents => 1.03288
+    0x010891a7, // +57 cents => 1.03347
+    0x0108b8ca, // +58 cents => 1.03407
+    0x0108dff1, // +59 cents => 1.03467
+    0x0109071f, // +60 cents => 1.03526
+    0x01092e53, // +61 cents => 1.03586
+    0x0109558c, // +62 cents => 1.03646
+    0x01097ccb, // +63 cents => 1.03706
+    0x0109a410, // +64 cents => 1.03766
+    0x0109cb5b, // +65 cents => 1.03826
+    0x0109f2ac, // +66 cents => 1.03886
+    0x010a1a02, // +67 cents => 1.03946
+    0x010a415e, // +68 cents => 1.04006
+    0x010a68c0, // +69 cents => 1.04066
+    0x010a9028, // +70 cents => 1.04126
+    0x010ab796, // +71 cents => 1.04186
+    0x010adf09, // +72 cents => 1.04247
+    0x010b0683, // +73 cents => 1.04307
+    0x010b2e02, // +74 cents => 1.04367
+    0x010b5587, // +75 cents => 1.04427
+    0x010b7d12, // +76 cents => 1.04488
+    0x010ba4a2, // +77 cents => 1.04548
+    0x010bcc39, // +78 cents => 1.04608
+    0x010bf3d5, // +79 cents => 1.04669
+    0x010c1b78, // +80 cents => 1.04729
+    0x010c4320, // +81 cents => 1.04790
+    0x010c6ace, // +82 cents => 1.04850
+    0x010c9282, // +83 cents => 1.04911
+    0x010cba3c, // +84 cents => 1.04972
+    0x010ce1fb, // +85 cents => 1.05032
+    0x010d09c1, // +86 cents => 1.05093
+    0x010d318c, // +87 cents => 1.05154
+    0x010d595d, // +88 cents => 1.05214
+    0x010d8135, // +89 cents => 1.05275
+    0x010da912, // +90 cents => 1.05336
+    0x010dd0f5, // +91 cents => 1.05397
+    0x010df8dd, // +92 cents => 1.05458
+    0x010e20cc, // +93 cents => 1.05519
+    0x010e48c1, // +94 cents => 1.05580
+    0x010e70bb, // +95 cents => 1.05641
+    0x010e98bc, // +96 cents => 1.05702
+    0x010ec0c2, // +97 cents => 1.05763
+    0x010ee8cf, // +98 cents => 1.05824
+    0x010f10e1, // +99 cents => 1.05885
+    0x010f38f9, // +100 cents => 1.05946
+};
+//==============================================================================
+// End generated code section
+//==============================================================================
+
+uq16_16 bendPitchWheel(uint8_t noteId, uint16_t pitchWheel)
+{
+    // First, convert the pitch wheel value to +/-cents
+    int32_t bendCents = (((int16_t)-0x2000) + pitchWheel) * 100 / 0x1FFF;
+
+    // Grab the base frequency and store it in a 64-bit int
+    uint64_t freq = noteFreqTable[noteId];
```

Comment:
64 bit math is much slower than 32 bit math on the ESP32-S2. Any way that could be avoided?

Reply:
Hmm, possibly. Let me see...

Right now the note frequencies are using 30 bits total of a uq16_16 but tbh the lowest 8 bits are probably not necessary. The pitch bend table is using 25 total bits of a uq8_24, that could also maybe live with dropping the lowest 8 bits.

But that still doesn't let us multiply into a uint32, so maybe instead... I could just shift the pitch bend value left by 6 or 7 bits and the frequency by 2 or 1. Then the result we want would be the top 32 bits of the multiply and in theory the compiler could use the `MULUH` instruction to get that value without any 64-bit math.

I should figure out how to check the assembly, maybe it's already doing something like that anyway...

Reply:
https://godbolt.org/ has Xtensa ESP32-S2 gcc 12.2.0, which is close enough. Just remember to give it the `-O2` option

Reply:
Alright, so I tried that and also objdump and I did what I said above and it went from:

```
  2f:   642992          l32i    a9, a9, 0x190
    uint64_t freq = noteFreqTable[noteId];
  32:   a08280          addx4   a8, a2, a8
    freq *= bendTable[bendCents + 100];
  35:   0888            l32i.n  a8, a8, 0
  37:   822980          mull    a2, a9, a8
  3a:   a29980          muluh   a9, a9, a8
    // Mask the result back into a 32-bit value
    uq16_16 trimmedFreq = ((freq >> 24) & (0xFFFFFFFFu));

    // Not using an intermediate variable here makes weird stuff happen
    return trimmedFreq;
}
  3d:   404810          ssai    24
  40:   812920          src     a2, a9, a2
  43:   f01d            retw.n
```
to

```
  34:   642922          l32i    a2, a9, 0x190
    uint64_t freq = noteFreqTable[noteId] << 2;
  37:   1188e0          slli    a8, a8, 2
    freq *= (bendTable[bendCents + 100] << 6);
  3a:   1122a0          slli    a2, a2, 6

    // Shift right 24 bits to account for the fractional component of the bend table
    // Mask the result back into a 32-bit value
    uq16_16 trimmedFreq = ((freq >> 32) & (0xFFFFFFFFu));
  3d:   a22280          muluh   a2, a2, a8
 
    // Not using an intermediate variable here makes weird stuff happen
    return trimmedFreq;
}
  40:   f01d            retw.n
```

Some of the shifts changed around but it basically turned `addx4, l32i, mull, muluh, ssai, src` into `slli, slli, muluh` - pretty nice! Perhaps I can improve the cents calculation a bit too...